cmd/data-crawler.go

```diff
@@ -402,19 +435,75 @@ func (f *folderScanner) scanQueuedLevels(ctx context.Context, folders []cachedFo
 		if err != nil {
 			return nil, err
 		}
+
+		if f.healObjectSelect == 0 {
+			// If we are not scanning, return now.
+			continue
+		}
+
+		// Whatever remains in existing must have been deleted or is missing and requires healing.
+		objAPI := newObjectLayerWithoutSafeModeFn()
+		for k := range existing {
+			f.waitForLowActiveIO()
+			bucket, prefix := path2BucketObject(k)
+			if f.dataUsageCrawlDebug {
+				logger.Info(color.Green("folder-scanner:")+" checking disappeared folder: %v/%v", bucket, prefix)
+			}
+			// Try as a folder first.
+			err := objAPI.HealObjects(ctx, bucket, prefix+slashSeparator, madmin.HealOpts{Recursive: true, Remove: healDeleteDangling},
+				func(bucket, object, versionID string) error {
+					_, err := objAPI.HealObject(ctx, bucket, object, versionID, madmin.HealOpts{Recursive: true, Remove: healDeleteDangling})
+					return err
+				})
+			if err != nil {
+				// If path is an object nil is still returned.
+				logger.LogIf(ctx, err)
+				continue
+			}
+
+			// If it still doesn't exist it may be an object.
+			_, err = os.Stat(filepath.Join(f.root, k))
+			if os.IsNotExist(err) {
+				// Try as an object.
+				_, err = objAPI.HealObject(ctx, bucket, prefix, "", madmin.HealOpts{Recursive: false, Remove: healDeleteDangling})
```

Comment:
If this is called on a data folder, it could see the data as dangling since there is no metadata. I will have to check that.

Reply:
You can still call HealObjects() to initiate a Walk() so that it only returns actual objects never a directory @klauspost